pkg/standalone-utils/test/PermitRelayer.test.ts

```diff
@@ -0,0 +1,432 @@
+
+  const tokenList = [await Token.fromContract(dai), await Token.fromContract(mkr), await Token.create('WETH')];
+  const tokens = new TokenList(tokenList.sort((a, b) => a.compare(b)));
+
+  // Deploy Balancer Vault
+  const authorizer = await deploy('v2-vault/Authorizer', { args: [admin.address] });
+  const vault = await deploy('v2-vault/Vault', { args: [authorizer.address, tokens.WETH.address, 0, 0] });
+
+  // Deploy Pool
+  const pool = await deploy('v2-vault/test/MockPool', { args: [vault.address, GeneralPool] });
+  const poolId = await pool.getPoolId();
+
+  await tokens.mint({ to: lp, amount: tokenInitialBalance.mul(2), from: admin });
+  await tokens.approve({ to: vault.address, from: [lp] });
+
+  await pool.registerTokens(tokens.addresses, [ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS]);
+
+  await vault.connect(lp).joinPool(poolId, lp.address, lp.address, {
+    assets: tokens.addresses,
+    maxAmountsIn: tokens.addresses.map(() => MAX_UINT256),
+    fromInternalBalance: false,
+    userData: encodeJoin(
+      tokens.addresses.map(() => tokenInitialBalance),
+      tokens.addresses.map(() => 0)
+    ),
+  });
+
+  await tokens.approve({ to: vault.address, from: [lp], amount: 0 });
+
+  const relayer = await deploy('PermitRelayer', { args: [vault.address] });
+
+  // Authorise relayer to perform actions.
+  const single = await actionId(vault, 'swap');
+  const batch = await actionId(vault, 'batchSwap');
+  const manageUserBalance = await actionId(vault, 'manageUserBalance');
+  const joinPool = await actionId(vault, 'joinPool');
+  const exitPool = await actionId(vault, 'exitPool');
+  await authorizer.connect(admin).grantRoles([single, batch, manageUserBalance, joinPool, exitPool], relayer.address);
+
+  return {
+    data: {
+      poolId,
+    },
+    contracts: {
+      tokens,
+      pool,
+      relayer,
+      vault,
+    },
+  };
+};
+
+describe('PermitRelayer', function () {
+  let tokens: TokenList, relayer: Contract, vault: Contract;
+
+  let lp: SignerWithAddress;
+  let poolId: string;
+
+  before('deploy base contracts', async () => {
+    [, , lp] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('set up asset manager', async () => {
+    const { data, contracts } = await setup();
+    poolId = data.poolId;
+
+    tokens = contracts.tokens;
+    relayer = contracts.relayer;
+    vault = contracts.vault;
+  });
+
+  describe('multicall', () => {
+    it('allows calling permit on multiple tokens', async () => {
+      const dai = tokens.DAI.instance;
+      const mkr = tokens.MKR.instance;
+
+      const daiSig = await signPermit(dai, lp, vault, MAX_UINT256);
+      const mkrSig = await signPermit(mkr, lp, vault, MAX_UINT256);
+
+      const daiPermit = relayer.interface.encodeFunctionData('vaultPermitDAI', [
+        dai.address,
+        daiSig.nonce,
+        MAX_UINT256,
+        true,
+        daiSig.v,
+        daiSig.r,
+        daiSig.s,
+      ]);
```

Comment:
I'd consider reducing repeated code and verboseness by e.g. extracting thise into the following:

```
context('with signed permit', () => {
  let daIPermit: string;
  sharedBeforeEach(async () => {
     daiPermit = await ...
  });
  
  it('can call permit on multiple tokens', async () => {
     const result = await multicall();
     expect(result).to.be.ok;
  })
})
```

It otherwise becomes harder to review each individual test, and how they are different from each other.